### DISCUSSION TITLE
Kornia vs OpenCV homography estimation and warping time

### DISCUSSION THREAD

**Discussion by mihaibujanca** [kornia/kornia#1566] [Questions] [answered]
Kornia version: '0.6.3'
PyTorch version: '1.9.0+cu102'
OpenCV version: '4.5.1'

Kornia (GPU) seems to be significantly slower than OpenCV (CPU) in computing the homography and warping an image, given a set of matched keypoints. Note that for Kornia, the default RANSAC max_iterations=10, while for OpenCV it's 2000, yet Kornia is way slower.

Is this a known issue? Am I doing this right?

```
# kornia vs opencv warp tests
from kornia.geometry.ransac import RANSAC
import kornia as K
import numpy as np
import torch
rows, cols = gray0.shape   # 480, 640
src_pts, dst_pts, confidences = feature_extractor.match(gray0, gray1)  # returns GPU tensors

def warp_cv(src_pts, dst_pts, frame):
    assert(type(frame) == np.ndarray)
    cpu_src_pts, cpu_dst_pts = src_pts.cpu().numpy(), dst_pts.cpu().numpy()  # take into account gpu -> cpu time
    
    T, mask = cv2.findHomography(cpu_src_pts, cpu_dst_pts)
    warped = cv2.warpPerspective(frame, T, (cols, rows))
    return warped


ransac = RANSAC('homography').cuda()
gpu_gray = K.image_to_tensor(gray0, keepdim=False).cuda()

def warp_kornia(src_pts, dst_pts, frame):
    assert(type(frame) == torch.Tensor)
    T, mask = ransac(src_pts, dst_pts)
    warped = K.geometry.transform.warp_perspective(frame.float(), T[None], (rows, cols))
    return warped

###  COMPUTE ###

%timeit -n 10 warp_kornia(src_pts, dst_pts, gpu_gray0)
> 105 ms ± 8.24 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit -n 10 warp_cv(src_pts, dst_pts, gray0)
> 1.13 ms ± 21.7 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by ducha-aiki** [comment, marked as answer]
Hi,

First kornia max_iterations is NOT 10, it is 10 * batch_size (default) = 10* 2048 = 20k (see the docs https://kornia.readthedocs.io/en/latest/geometry.ransac.html#kornia.geometry.ransac.RANSAC )
 while for OpenCV it is 2k.
 
Second, it is true that OpenCV RANSAC is much more polished (also, RANSAC is sequential algorithm) and often stops after 10 or 50 iterations. 

Third, we do plan to improve kornia RANSAC, however it is not straightforward from the research and implementation point of view, also we all are doing kornia is our free time, whereas OpenCV has full-time developers. 
So consider sponsoring the kornia with pull requests or money ;)

**Comment by mihaibujanca** [comment]
Hello,

Thanks very much for the quick answer and your work on kornia!
I should've paid more attention to the docs wrt the number of iterations.
I've just started exploring kornia and hope to contribute in the future.
